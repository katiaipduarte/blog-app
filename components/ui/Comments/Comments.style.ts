import styled from 'styled-components';

export const CommentsContainer = styled.section`
  margin: 2rem 0;
  text-align: justify;

  ul {
    margin: 2rem 0;

    li {
      margin: 1rem 0;

      h3 {
        color: ${({ theme }) => theme.grey};
        font-size: 0.75rem;
      }

      p {
        margin-top: 0.25rem;
        font-size: 0.5rem;
      }

      .parent,
      .child {
        display: flex;
        justify-content: space-between;
      }

      .child {
        width: calc(100% - 1.5rem);
        paddding: 2rem 0;
        margin: 1rem 0 1rem 1.5rem;
      }
    }
  }

  @media all and (min-width: 960px) {
    ul {
      li {
        h3 {
          font-size: 0.8rem;
        }

        p {
          margin-top: 0.25rem;
          font-size: 1rem;
        }

        .child {
          max-width: 30rem;
          paddding: 2rem 0;
          margin: 1rem 0 1rem 3.5rem;
        }
      }
    }
  }
`;
