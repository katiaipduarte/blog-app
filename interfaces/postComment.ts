export type PostComment = {
  content: string;
  date: string;
  id: number;
  parent_id: number | null;
  postId: number;
  user: string;
};
