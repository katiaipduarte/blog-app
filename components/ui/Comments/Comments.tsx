import { PostComment } from '@interfaces/postComment';
import { useFetch } from '@lib/hooks/useFetch';
import React, { memo, useState } from 'react';
import CommentForm from '../CommentForm/CommentForm';
import { CommentsContainer } from './Comments.style';

type Props = {
  comments: PostComment[];
  postId: number;
};

const Comments = (props: Props): JSX.Element => {
  const { comments, postId } = props;
  const [parentId, setParentId] = useState<number | null>(null);
  const [parentUser, setParentUser] = useState<string>('');
  const [commentList, setCommentList] = useState<PostComment[]>(comments);

  const renderComments = (): JSX.Element => {
    return (
      <ul>
        {commentList.map((comment: PostComment) => (
          <li key={comment.id}>
            {!comment.parent_id && buildParentComments(comment)}
            {buildChildrenComments(comment.id, comment.user)}
          </li>
        ))}
      </ul>
    );
  };

  const buildParentComments = (comment: PostComment): JSX.Element => {
    return (
      <div className="parent">
        <div>
          <h4>{`${comment.user} on ${comment.date}`}</h4>
          <p>{comment.content}</p>
        </div>
        <button
          role="button"
          className="full"
          onClick={() => handleReply(comment.id, comment.user)}
          onKeyPress={() => handleReply(comment.id, comment.user)}
        >
          Reply
        </button>
      </div>
    );
  };

  const buildChildrenComments = (id: number, user: string): JSX.Element => {
    return (
      <>
        {comments
          .filter((i: PostComment) => i.parent_id === id)
          .map((comment: PostComment) => {
            return (
              <div className="child" key={comment.id}>
                <div>
                  <h4>{`${comment.user} response to ${user} on ${comment.date}`}</h4>
                  <p>{comment.content}</p>
                </div>

                <button
                  role="button"
                  className="full"
                  onClick={() => handleReply(comment.id, comment.user)}
                  onKeyPress={() => handleReply(comment.id, comment.user)}
                >
                  Reply
                </button>
              </div>
            );
          })}
      </>
    );
  };

  const handleReply = (parentCommentId: number, user: string): void => {
    setParentId(parentCommentId);
    setParentUser(user);
  };

  const commentAdded = async (): Promise<void> => {
    setParentId(null);
    setParentUser('');

    const commentsReq = await fetch(
      `${process.env.NEXT_PUBLIC_API_ENDPOINT}/posts/${postId}/comments`
    );
    const comments = await commentsReq.json();
    setCommentList(comments);
  };

  return (
    <CommentsContainer>
      <h2>{`${comments.length} Comments`}</h2>
      {renderComments()}
      <CommentForm
        postId={postId}
        parentCommentId={parentId}
        user={parentUser}
        commentAdded={commentAdded}
      />
    </CommentsContainer>
  );
};

export default memo(Comments);
