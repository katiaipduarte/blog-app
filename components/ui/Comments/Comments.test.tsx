import { PostComment } from '@interfaces/postComment';
import 'jest-styled-components';
import React from 'react';
import renderer from 'react-test-renderer';
import Comments from './Comments';

const mockedComments: PostComment[] = [
  {
    content: 'banana',
    user: 'test',
    date: '2021-06-28',
    parent_id: null,
    postId: 10,
    id: 8
  },
  {
    content: 'bananan 2x',
    user: 'yolo',
    date: '2021-06-28',
    parent_id: 8,
    postId: 10,
    id: 9
  }
];

describe('the <Comments /> component', () => {
  it('should render comments list', () => {
    const tree = renderer
      .create(<Comments postId={10} comments={mockedComments} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render 0 comments', () => {
    const tree = renderer
      .create(<Comments postId={10} comments={[] as PostComment[]} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
