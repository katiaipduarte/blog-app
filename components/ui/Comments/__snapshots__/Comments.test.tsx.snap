// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the <Comments /> component should render 0 comments 1`] = `
.c1 {
  margin: 2rem auto;
}

.c1 p {
  font-size: 1.25rem;
  line-height: 1.75rem;
}

.c1 div {
  margin: 2rem 0;
}

.c1 div input[type='text'],
.c1 div textarea {
  width: 100% !important;
  border-radius: 0.4rem;
  box-sizing: border-box;
  margin: 1rem 0;
  box-shadow: 0 0 20px 0 rgba(0,0,0,0.2),0 5px 5px 0 rgba(0,0,0,0.24);
  padding: 0.5rem;
  display: block;
}

.c1 div input[type='text']:focus-visible,
.c1 div textarea:focus-visible,
.c1 div input[type='text']:focus:not(:focus-visible),
.c1 div textarea:focus:not(:focus-visible) {
  outline: none;
  border: 2px solid;
}

.c1 div input[type='text'] {
  height: 2rem;
}

.c1 div textarea {
  height: 10rem !important;
  resize: none;
}

.c1 div .invalid {
  border: 2px solid !important;
}

.c1 div .message {
  font-size: 0.75rem;
  display: none;
}

.c1 div .show {
  display: block !important;
}

.c0 {
  margin: 2rem 0;
  text-align: justify;
}

.c0 ul {
  margin: 2rem 0;
}

.c0 ul li {
  margin: 1rem 0;
}

.c0 ul li h4 {
  font-size: 0.75rem;
}

.c0 ul li p {
  margin-top: 0.25rem;
  font-size: 0.5rem;
}

.c0 ul li .parent,
.c0 ul li .child {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c0 ul li .child {
  width: calc(100% - 1.5rem);
  paddding: 2rem 0;
  margin: 1rem 0 1rem 1.5rem;
}

@media all and (min-width:960px) {
  .c1 {
    text-align: left;
  }

  .c1 div input[type='text'],
  .c1 div textarea {
    width: 70% !important;
  }
}

@media all and (min-width:960px) {
  .c0 ul li h4 {
    font-size: 0.8rem;
  }

  .c0 ul li p {
    margin-top: 0.25rem;
    font-size: 1rem;
  }

  .c0 ul li .child {
    max-width: 30rem;
    paddding: 2rem 0;
    margin: 1rem 0 1rem 3.5rem;
  }
}

<section
  className="c0"
>
  <h2
    role="text"
    tabIndex={0}
  >
    0 Comments
  </h2>
  <ul />
  <form
    className="c1"
    onSubmit={[Function]}
  >
    <h2>
      Submit a comment
    </h2>
    
    <div>
      <input
        className=""
        id="user"
        onChange={[Function]}
        placeholder="Username"
        type="text"
        value=""
      />
      <span
        className="message"
      >
        Username cannot be empty
      </span>
    </div>
    <div>
      <textarea
        className=""
        id="content"
        name="comment"
        onChange={[Function]}
        placeholder="Write your comment here..."
        value=""
      />
      <span
        className="message"
      >
        Comment cannot be empty
      </span>
    </div>
    <button
      className="full"
      type="submit"
    >
      Submit
    </button>
  </form>
</section>
`;

exports[`the <Comments /> component should render comments list 1`] = `
.c1 {
  margin: 2rem auto;
}

.c1 p {
  font-size: 1.25rem;
  line-height: 1.75rem;
}

.c1 div {
  margin: 2rem 0;
}

.c1 div input[type='text'],
.c1 div textarea {
  width: 100% !important;
  border-radius: 0.4rem;
  box-sizing: border-box;
  margin: 1rem 0;
  box-shadow: 0 0 20px 0 rgba(0,0,0,0.2),0 5px 5px 0 rgba(0,0,0,0.24);
  padding: 0.5rem;
  display: block;
}

.c1 div input[type='text']:focus-visible,
.c1 div textarea:focus-visible,
.c1 div input[type='text']:focus:not(:focus-visible),
.c1 div textarea:focus:not(:focus-visible) {
  outline: none;
  border: 2px solid;
}

.c1 div input[type='text'] {
  height: 2rem;
}

.c1 div textarea {
  height: 10rem !important;
  resize: none;
}

.c1 div .invalid {
  border: 2px solid !important;
}

.c1 div .message {
  font-size: 0.75rem;
  display: none;
}

.c1 div .show {
  display: block !important;
}

.c0 {
  margin: 2rem 0;
  text-align: justify;
}

.c0 ul {
  margin: 2rem 0;
}

.c0 ul li {
  margin: 1rem 0;
}

.c0 ul li h4 {
  font-size: 0.75rem;
}

.c0 ul li p {
  margin-top: 0.25rem;
  font-size: 0.5rem;
}

.c0 ul li .parent,
.c0 ul li .child {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c0 ul li .child {
  width: calc(100% - 1.5rem);
  paddding: 2rem 0;
  margin: 1rem 0 1rem 1.5rem;
}

@media all and (min-width:960px) {
  .c1 {
    text-align: left;
  }

  .c1 div input[type='text'],
  .c1 div textarea {
    width: 70% !important;
  }
}

@media all and (min-width:960px) {
  .c0 ul li h4 {
    font-size: 0.8rem;
  }

  .c0 ul li p {
    margin-top: 0.25rem;
    font-size: 1rem;
  }

  .c0 ul li .child {
    max-width: 30rem;
    paddding: 2rem 0;
    margin: 1rem 0 1rem 3.5rem;
  }
}

<section
  className="c0"
>
  <h2
    role="text"
    tabIndex={0}
  >
    2 Comments
  </h2>
  <ul>
    <li>
      <div
        className="parent"
      >
        <div>
          <h4
            tabIndex={0}
          >
            test on 2021-06-28
          </h4>
          <p
            tabIndex={0}
          >
            banana
          </p>
        </div>
        <button
          className="full"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="button"
        >
          Reply
        </button>
      </div>
      <div
        className="child"
      >
        <div>
          <h4
            role="text"
            tabIndex={0}
          >
            yolo response to test on 2021-06-28
          </h4>
          <p
            role="text"
            tabIndex={0}
          >
            bananan 2x
          </p>
        </div>
      </div>
    </li>
    <li />
  </ul>
  <form
    className="c1"
    onSubmit={[Function]}
  >
    <h2>
      Submit a comment
    </h2>
    
    <div>
      <input
        className=""
        id="user"
        onChange={[Function]}
        placeholder="Username"
        type="text"
        value=""
      />
      <span
        className="message"
      >
        Username cannot be empty
      </span>
    </div>
    <div>
      <textarea
        className=""
        id="content"
        name="comment"
        onChange={[Function]}
        placeholder="Write your comment here..."
        value=""
      />
      <span
        className="message"
      >
        Comment cannot be empty
      </span>
    </div>
    <button
      className="full"
      type="submit"
    >
      Submit
    </button>
  </form>
</section>
`;
